airframe-control/src/main/scala/wvlet/airframe/control/CircuitBreaker.scala

```diff
@@ -248,34 +306,50 @@ case class CircuitBreaker(
     * Execute the body block through the CircuitBreaker.
     *
     * If the state is OPEN, this will throw CircuitBreakerOpenException (fail-fast). The state will move to HALF_OPEN state
-    * after a cetain amount of delay, determined by the delayAfterMarkedDead policy.
+    * after a certain amount of delay, determined by the delayAfterMarkedDead policy.
     *
     * If the state is HALF_OPEN, this method allows running the code block once, and if the result is successful,
     * the state will move to CLOSED. If not, the state will be OPEN again.
     *
     * If the state is CLOSED, the code block will be executed normally. If the result is marked failure or nonRetryable exception
-    * is thrown, it will report to the failure to the HealthCheckPolicy. If this policy determins the target service is dead,
+    * is thrown, it will report to the failure to the HealthCheckPolicy. If this policy determines the target service is dead,
     * the circuit will shift to OPEN state to block the future execution.
     *
     */
-  def run[A](body: => A): Unit = {
+  def run[A: ClassTag](body: => A): A = {
     verifyConnection
 
     val result = Try(body)
+
     val resultClass = result match {
-      case Success(x) => resultClassifier(x)
-      case Failure(RetryableFailure(e)) =>
-        ResultClass.retryableFailure(e)
-      case Failure(e) =>
-        errorClassifier(e)
+      case Success(x)                   => resultClassifier(x)
+      case Failure(RetryableFailure(e)) => ResultClass.retryableFailure(e)
+      case Failure(e)                   => errorClassifier(e)
     }
+
     resultClass match {
-      case Succeeded =>
+      case ResultClass.Succeeded(x) =>
```

Comment:
I'm against embedding the result object here as we should have it in result variable. Why can't we use only failbackHandler?

Reply:
This is because of errorClassifier can convert `Failure` to `Succeeded` and it causes `RuntimeError` at https://github.com/wvlet/airframe/blob/dafedf03e500ad140cfab28f3dd76388caa8f6bb/airframe-control/src/main/scala/wvlet/airframe/control/CircuitBreaker.scala#L275

I once thought that modifying errorClassifier to be allowed only classifying retryable or not here, but the same thing happens in `Retry`.
https://github.com/wvlet/airframe/blob/dafedf03e500ad140cfab28f3dd76388caa8f6bb/airframe-control/src/main/scala/wvlet/airframe/control/Retry.scala#L284

I would say the current structure is broken. We need to change something to fix anyway but I'm not sure what's the best solution covering both `CircuitBreaker` and `Retry`.

> Why can't we use only failbackHandler?

How fallbackHandler can help this situation?

Reply:
Maybe errorClassifier should be `Throwable => ResultClass.Failed`?

Reply:
What I think is:
• ResultClassifier should not generate any result, so converting Failure to Succeeded should never happen. 
• When we see Failed(retryable/ignorable = true), we just throw this exception so that Retry can submit the request again.
• When we see Failed(retryable = false), we should apply a fallback handler to produce some fallback results. The default behavior is just throwing an exception. If the user provides a custom default result, we should return that by casting it to A.

Reply:
> ResultClassifier should not generate any result, so converting Failure to Succeeded should never happen. Try.Failure can be classified only for Failed(retryable = true or false)

Just confirmation. We have two types of classifiers `resultClassifier` and `errorClassifier`. I'd like to clarify the role of them.

- `errorClassifier` is responsible for determining if error is ignoreable(= retryable). He can not convert failure as success.
- On the other hand, still `resultClassifier` can convert success as failure?

> Question: If the user really needs to classify some Failure response as Succeeded, should we apply the fallback handler to produce the default result?

Not sure, but isolating the classifying result logic and the producing default value logic might make the situation hard if we need to produce the default value depending on the error type.